### PR TITLE
display selected gateway clients 

### PR DESCRIPTION
**Pull request by Idadelveloper**
The selected gateway client and other info will be displayed on the screen and would change if the user selects another gateway client.

### CODE REVIEW

File: app/src/main/java/com/example/sw0b_001/Homepage/GatewayClientListingFragment.kt [hunk 1]
```diff
@@ -130,7 +148,8 @@ class GatewayClientListingFragment : Fragment(R.layout.activity_gateway_clients_
     }
 
     class GatewayClientListingAdapter(private val gatewayClientsCommunications: GatewayClientsCommunications,
```
Comment (sherlockwisdom) [inline app/src/main/java/com/example/sw0b_001/Homepage/GatewayClientListingFragment.kt:150]: @Idadelveloper You shouldn't pass viewModel to the Adapter, they always have to be independent of each other

File: app/src/main/java/com/example/sw0b_001/Models/GatewayClients/GatewayClientViewModel.kt [hunk 2]
```diff
@@ -20,17 +26,32 @@ class GatewayClientViewModel : ViewModel() {
         return liveData
     }
 
+    fun updateSelectedGatewayClient(newGatewayClientMsisdn: String?) {
```
Comment (sherlockwisdom) [inline app/src/main/java/com/example/sw0b_001/Models/GatewayClients/GatewayClientViewModel.kt:29]: No need for this

File: app/src/main/java/com/example/sw0b_001/Models/GatewayClients/GatewayClientViewModel.kt [hunk 3]
```diff
@@ -8,10 +8,16 @@ import androidx.lifecycle.ViewModel
...
 import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.flow.MutableStateFlow
+import kotlinx.coroutines.flow.StateFlow
+import kotlinx.coroutines.flow.asStateFlow
 import kotlinx.coroutines.launch
 
-class GatewayClientViewModel : ViewModel() {
+class GatewayClientViewModel() : ViewModel() {
     private var liveData: LiveData<List<GatewayClient>> = MutableLiveData()
+    private val _selectedGatewayClient = MutableStateFlow<GatewayClient?>(null)
```
Comment (sherlockwisdom) [inline app/src/main/java/com/example/sw0b_001/Models/GatewayClients/GatewayClientViewModel.kt:18]: No need for this

File: app/src/main/java/com/example/sw0b_001/Models/GatewayClients/GatewayClientViewModel.kt [hunk 4]
```diff
@@ -20,17 +26,32 @@ class GatewayClientViewModel : ViewModel() {
...
+                _selectedGatewayClient.value = defaultGatewayClient
                 successRunnable?.run()
             } catch (e: Exception) {
                 Log.e(javaClass.name, "Exception fetching Gateway clients", e)
                 failureRunnable?.run()
             }
         }
     }
+
+    private fun getGatewayClientByMsisdn(context: Context, msisdn: String): GatewayClient? {
```
Comment (sherlockwisdom) [inline app/src/main/java/com/example/sw0b_001/Models/GatewayClients/GatewayClientViewModel.kt:54]: This is good, look for the function below in GatewayCLientListingFragment and use the sharedpreference change to fetch from getGatewayClientByMsisdn and make the necessary changes 

``` 
private val sharedPreferencesChangeListener = OnSharedPreferenceChangeListener { _, _ ->
        if(::listViewAdapter.isInitialized)
            listViewAdapter.notifyDataSetChanged()
    }
   ```

File: app/src/main/java/com/example/sw0b_001/Homepage/GatewayClientListingFragment.kt [hunk 5]
```diff
@@ -157,27 +176,28 @@ class GatewayClientListingFragment : Fragment(R.layout.activity_gateway_clients_
...
-        private fun gatewayClientOnClickListener(gatewayClient: GatewayClient):
+        private fun gatewayClientOnClickListener(gatewayClient: GatewayClient, viewModel: GatewayClientViewModel):
                 OnClickListener {
             return OnClickListener {
                 gatewayClientsCommunications.updateDefaultGatewayClient(gatewayClient.mSISDN!!)
                 CoroutineScope(Dispatchers.Default).launch{
                     gatewayClient.type = "custom"
                     Datastore.getDatastore(it.context).gatewayClientsDao().update(gatewayClient)
                 }
+                viewModel.updateSelectedGatewayClient(gatewayClient.mSISDN)
```
Comment (sherlockwisdom) [inline app/src/main/java/com/example/sw0b_001/Homepage/GatewayClientListingFragment.kt:200]: You can call methods to update UI from here

**Review by sherlockwisdom — CHANGES_REQUESTED** [review]

File: app/src/main/java/com/example/sw0b_001/Homepage/GatewayClientListingFragment.kt (hunk 5 above)
Comment (Idadelveloper) [reply app/src/main/java/com/example/sw0b_001/Homepage/GatewayClientListingFragment.kt]: @sherlockwisdom I just made the changes

**Pull request merged**